### PR TITLE
etcdserver: reject v3 txns with duplicate put keys

### PR DESCRIPTION
**Pull request by heyitsanthony**
An API check to support PR #4363; bad requests didn't return an error.

### CODE REVIEW

File: etcdserver/api/v3rpc/key.go [hunk 1]
```diff
@@ -176,13 +176,40 @@ func checkTxnRequest(r *pb.TxnRequest) error {
...
 	}
+	if err := checkRequestPutKeys(r.Failure); err != nil {
+		return err
+	}
+
+	return nil
+}
 
+// checkRequestPutKeys gives ErrDuplicateKey if the same key is put twice
+func checkRequestPutKeys(reqs []*pb.RequestUnion) error {
```
Comment (xiang90) [inline etcdserver/api/v3rpc/key.go]: we should not allow put then delete too, right?

File: integration/v3_grpc_test.go [hunk 2]
```diff
@@ -132,6 +139,19 @@ func TestV3TxnTooManyOps(t *testing.T) {
 			t.Errorf("#%d: err = %v, want %v", i, err, v3rpc.ErrTooManyOps)
 		}
 	}
+
+	// equal keys fail earlier
+	keyf = func() []byte { return []byte("key") }
```
Comment (xiang90) [inline integration/v3_grpc_test.go]: probably we should create a new test. TestV3TxnDuplicateKeys?

File: etcdserver/api/v3rpc/key.go (hunk 1 above)
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/key.go]: Right, will fix

**Comment by xiang90** [comment]
Cannot sleep early recently for no reason. etcd night club formed.

File: integration/v3_grpc_test.go (hunk 2 above)
Comment (heyitsanthony) [inline integration/v3_grpc_test.go]: OK

**Comment by heyitsanthony** [comment]
all fixed ptal /cc @xiang90

**Comment by xiang90** [comment]
LGTM

**Comment by heyitsanthony** [comment]
ok merging on greenlight

**Pull request merged**